h2o-automl/src/main/java/ai/h2o/automl/targetencoding/TargetEncoder.java

```diff
@@ -0,0 +1,745 @@
+     * @param foldColumnName name of the column that contains fold number the row is belong to
+     * @param inputeNAsWithNewCategory set to `true` to impute NAs with new category.     // TODO probably we need to always set it to true bc we do not support null values on the right side of merge operation.
+     */
+    //TODO do we need to do this preparation before as a separate phase? because we are grouping twice.
+    //TODO At least it seems that way in the case of KFold. But even if we need to preprocess for other types of TE calculations... we should not affect KFOLD case anyway.
+    public Map<String, Frame> prepareEncodingMap(Frame data,
+                                                 String targetColumnName,
+                                                 String foldColumnName,
+                                                 boolean inputeNAsWithNewCategory) {
+
+        // Validate input data. Not sure whether we should check some of these.
+        // It will become clear when we decide if TE is going to be exposed to user or only integrated into AutoML's pipeline
+
+        if(data == null) throw new IllegalStateException("Argument 'data' is missing, with no default");
+
+        if(targetColumnName == null || targetColumnName.equals(""))
+            throw new IllegalStateException("Argument 'target' is missing, with no default");
+
+        if(! checkAllTEColumnsAreCategorical(data, columnNamesToEncode))
+            throw new IllegalStateException("Argument 'columnsToEncode' should contain only names of categorical columns");
+
+        if(Arrays.asList(columnNamesToEncode).contains(targetColumnName)) {
+            throw new IllegalArgumentException("Columns for target encoding contain target column.");
+        }
+
+        int targetIndex = getColumnIndexByName(data, targetColumnName);
+
+        //TODO Losing data here, we should use clustering to assign instances with some reasonable target values.
+        Frame  dataWithoutNAsForTarget = filterOutNAsFromTargetColumn(data, targetIndex);
+
+        Frame dataWithEncodedTarget = ensureTargetColumnIsNumericOrBinaryCategorical(dataWithoutNAsForTarget, targetIndex);
+
+        Map<String, Frame> columnToEncodingMap = new HashMap<String, Frame>();
+
+        for ( String teColumnName: columnNamesToEncode) { // TODO maybe we can do it in parallel
+            Frame teColumnFrame = null;
+
+            if(inputeNAsWithNewCategory) {
+              imputeNAsForColumn(dataWithEncodedTarget, teColumnName, teColumnName + "_NA");
+            }
+
+            teColumnFrame = groupThenAggregateForNumeratorAndDenominator(dataWithEncodedTarget, teColumnName, foldColumnName, targetIndex);
+
+            teColumnFrame.renameColumn( "sum_" + targetColumnName, "numerator");
+            teColumnFrame.renameColumn( "nrow", "denominator");
+
+            columnToEncodingMap.put(teColumnName, teColumnFrame);
+        }
+
+        dataWithEncodedTarget.delete();
```

Comment:
we should handle exceptional case where the method exist before the cleanup because of an exception